In [3]:
pip install boto3

  Using cached jmespath-1.0.1-py3-none-any.whl.metadata (7.6 kB)
  Using cached s3transfer-0.11.2-py3-none-any.whl.metadata (1.7 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.2/139.2 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.3/13.3 MB 75.6 MB/s eta 0:00:00:00:0100:01
Using cached jmespath-1.0.1-py3-none-any.whl (20 kB)
Using cached s3transfer-0.11.2-py3-none-any.whl (84 kB)
Note: you may need to restart the kernel to use updated packages.


In [7]:
# Loading required package

import boto3
from io import BytesIO
import pandas as pd
# you need to change the credentials for yourself
# Note that aws_access_key_id changes from time to time

s3 = boto3.client('s3',
                  aws_access_key_id='ASIAYAAO5HRMGMR6VLZG',
                  aws_secret_access_key='6kXSuniH1tE8XQDZortpmWPQqaUMhKC0WDgc9as1',
                  aws_session_token='IQoJb3JpZ2luX2VjELL//////////wEaCXVzLWVhc3QtMiJGMEQCIBDmD+Pm4X75Q62yHc22OhaOezPh9CWKzc4swGfjhFxwAiAK9YDgC44yLjFDVdC7gx4wc34EsDf2qWnIDbs9TQwRDyr0AgjL//////////8BEAAaDDU0OTc4NzA5MDAwOCIMJ5I5/CAlMXQx/UXIKsgCquc02CyEB8ABbd7EohtIeui4OXZycQDY+peBRj7ZQrYVJ8+5Z4u5AnUWJAH+iXXyZAhVq+pzixQJJJK4hfA98jk+MVM5mgN4NjvfTqwW5x5vXJH58MYyvE7l1ZtoNja47wAl7DxTnsKEkBcmw0diEOrTuz7MRbOXlzcCae9kqqJyhD2Ji9I17z/PK+UCJllXlWHU5CsivfMHwhPmvY+ZEeq0VwOmUadaQyfNigFOmSUja7SRZ1L9whuvaOdFOX/PA4TG4Si4P48AclpLx2YUThLTi0S0LWkUeoQrpCmPUT46mOb8iciarLkbeUrctjPc/viOI8uqPudnZ5aCWYYeIRDIhwgTLkQTYSh4DBr0Ki+ZWrCSPZOKtwBC/IeeNpblASyB56U+f6cGK2SJH/6pjInONQGOwqCEBZ7EXRbpcGFQfL5JAYIGOTCa5Kq9BjqoAWYZUU1O1EvpQnf5xl/tbcJ1sN9QSjHpv/XzFlxzXI9CyaMhTPcaWtzjPzb32LvQ/LOKR6YzXZRfx3Dqux8gcdlIhTk+r5aY0ldG55jMgz6SHD4GiuFaBXWZyYkSccAo5BQMxoOajV5uEgb90n04yiY+8EeDr0aaPnxSuIJyEmH30L8S94TUtDu18bDYLK3ltvFoRZR5Uo2VJVcKObUwoLPzK0r3WwUaoA==')


bucket_name = 'de300winter2025'
object_key = 'desmond_nebah/heart_disease.csv'
csv_obj = s3.get_object(Bucket=bucket_name, Key=object_key)
body = csv_obj['Body']
csv_string = body.read().decode('utf-8')
df = pd.read_csv(BytesIO(csv_string.encode()))
print(df.head())

  age  sex  painloc  painexer  relrest  pncaden   cp  trestbps  htn   chol  \
0  63  1.0      NaN       NaN      NaN      NaN  1.0     145.0  1.0  233.0   
1  67  1.0      NaN       NaN      NaN      NaN  4.0     160.0  1.0  286.0   
2  67  1.0      NaN       NaN      NaN      NaN  4.0     120.0  1.0  229.0   
3  37  1.0      NaN       NaN      NaN      NaN  3.0     130.0  0.0  250.0   
4  41  0.0      NaN       NaN      NaN      NaN  2.0     130.0  1.0  204.0   

   ...  exeref  exerwm  thal  thalsev  thalpul  earlobe  cmo  cday   cyr  \
0  ...     NaN     NaN   6.0      NaN      NaN      NaN  2.0  16.0  81.0   
1  ...     NaN     NaN   3.0      NaN      NaN      NaN  2.0   5.0  81.0   
2  ...     NaN     NaN   7.0      NaN      NaN      NaN  2.0  20.0  81.0   
3  ...     NaN     NaN   3.0      NaN      NaN      NaN  2.0   4.0  81.0   
4  ...     NaN     NaN   3.0      NaN      NaN      NaN  2.0  18.0  81.0   

   target  
0     0.0  
1     1.0  
2     1.0  
3     0.0  
4     0.0  

[

In [23]:
# Cleaning the Data

# Cleaning Step 1

# List of columns to retain
columns_to_retain = [
    'age', 'sex', 'painloc', 'painexer', 'cp', 'trestbps', 'smoke', 'fbs',
    'prop', 'nitr', 'pro', 'diuretic', 'thaldur', 'thalach', 'exang', 'oldpeak', 'slope']

# Retaining the specified columns
df_cleaned = df[columns_to_retain]

print(df_cleaned.head())

  age  sex  painloc  painexer   cp  trestbps  smoke  fbs  prop  nitr  pro  \
0  63  1.0      NaN       NaN  1.0     145.0    NaN  1.0   0.0   0.0  0.0   
1  67  1.0      NaN       NaN  4.0     160.0    NaN  0.0   1.0   0.0  0.0   
2  67  1.0      NaN       NaN  4.0     120.0    NaN  0.0   1.0   0.0  0.0   
3  37  1.0      NaN       NaN  3.0     130.0    NaN  0.0   1.0   0.0  0.0   
4  41  0.0      NaN       NaN  2.0     130.0    NaN  0.0   0.0   0.0  0.0   

   diuretic  thaldur  thalach  exang  oldpeak  slope  
0       0.0     10.5    150.0    0.0      2.3    3.0  
1       0.0      9.5    108.0    1.0      1.5    2.0  
2       0.0      8.5    129.0    1.0      2.6    2.0  
3       0.0     13.0    187.0    0.0      3.5    3.0  
4       0.0      7.0    172.0    0.0      1.4    1.0  


In [24]:
# Cleaning Step 2

# Creating a copy of the dataframe to avoid SettingWithCopyWarning
df_cleaned = df[columns_to_retain].copy()

# Step 1: Impute missing values for 'painloc' and 'painexer'
df_cleaned.loc[:, 'painloc'] = df_cleaned['painloc'].fillna(df_cleaned['painloc'].mode()[0])
df_cleaned.loc[:, 'painexer'] = df_cleaned['painexer'].fillna(df_cleaned['painexer'].mode()[0])

# Step 2: Replace values less than 100 mm Hg for 'trestbps'
df_cleaned.loc[:, 'trestbps'] = df_cleaned['trestbps'].apply(lambda x: x if x >= 100 else df_cleaned['trestbps'].median())

# Step 3: Replace values less than 0 and greater than 4 for 'oldpeak'
df_cleaned.loc[:, 'oldpeak'] = df_cleaned['oldpeak'].apply(lambda x: x if 0 <= x <= 4 else df_cleaned['oldpeak'].median())

# Step 4: Impute missing values for 'thaldur' and 'thalach'
df_cleaned.loc[:, 'thaldur'] = df_cleaned['thaldur'].fillna(df_cleaned['thaldur'].median())
df_cleaned.loc[:, 'thalach'] = df_cleaned['thalach'].fillna(df_cleaned['thalach'].median())

# Step 5: Replace missing values and values greater than 1 for 'fbs', 'prop', 'nitr', 'pro', 'diuretic'
for col in ['fbs', 'prop', 'nitr', 'pro', 'diuretic']:
    df_cleaned.loc[:, col] = df_cleaned[col].fillna(df_cleaned[col].mode()[0])
    df_cleaned.loc[:, col] = df_cleaned[col].apply(lambda x: 1 if x > 1 else x)

# Step 6: Impute missing values for 'exang' and 'slope'
df_cleaned.loc[:, 'exang'] = df_cleaned['exang'].fillna(df_cleaned['exang'].mode()[0])
df_cleaned.loc[:, 'slope'] = df_cleaned['slope'].fillna(df_cleaned['slope'].mode()[0])


print(df_cleaned.head())

  age  sex  painloc  painexer   cp  trestbps  smoke  fbs  prop  nitr  pro  \
0  63  1.0      1.0       1.0  1.0     145.0    NaN  1.0   0.0   0.0  0.0   
1  67  1.0      1.0       1.0  4.0     160.0    NaN  0.0   1.0   0.0  0.0   
2  67  1.0      1.0       1.0  4.0     120.0    NaN  0.0   1.0   0.0  0.0   
3  37  1.0      1.0       1.0  3.0     130.0    NaN  0.0   1.0   0.0  0.0   
4  41  0.0      1.0       1.0  2.0     130.0    NaN  0.0   0.0   0.0  0.0   

   diuretic  thaldur  thalach  exang  oldpeak  slope  
0       0.0     10.5    150.0    0.0      2.3    3.0  
1       0.0      9.5    108.0    1.0      1.5    2.0  
2       0.0      8.5    129.0    1.0      2.6    2.0  
3       0.0     13.0    187.0    0.0      3.5    3.0  
4       0.0      7.0    172.0    0.0      1.4    1.0  


Cleaning Step 3

For Source 1, I used the "Proportion of people 15 years and over who were current daily smokers by age and sex, 2022" dataset because it allows me to impute smoking rates with a higher level of precision by considering both age and sex, which will likely yield more accurate imputation for missing values in the "smoke" column given that the heart disease dataset also has both age and sex information.

For source 2, I used the tobacco product use - cigarettes dataset and assumed that the smoking rates by different age groups for both men and women are the same.


In [26]:
# Updated smoking rates for Source 1 and Source 2
smoking_rates_source1 = {
    '15–17': {'Male': 1.2, 'Female': 1.8},
    '18–24': {'Male': 9.3, 'Female': 5.9},
    '25–34': {'Male': 13.4, 'Female': 8.8},
    '35–44': {'Male': 13.5, 'Female': 8.5},
    '45–54': {'Male': 15.3, 'Female': 11.6},
    '55–64': {'Male': 17.4, 'Female': 12.0},
    '65–74': {'Male': 9.9, 'Female': 7.9},
    '75+': {'Male': 3.8, 'Female': 1.9}
}

smoking_rates_source2 = {
    '18–24': {'Male': 4.8, 'Female': 4.8},
    '25–44': {'Male': 12.5, 'Female': 12.5},
    '45–64': {'Male': 15.1, 'Female': 15.1},
    '65+': {'Male': 8.7, 'Female': 8.7}
}

source2_smoking_rate_among_men = 13.2
source2_smoking_rate_among_women = 10.0

# Convert the 'age' column to numeric, coercing any non-numeric values to NaN
df_cleaned['age'] = pd.to_numeric(df_cleaned['age'], errors='coerce')

# For Source 1, simply replace missing values with the corresponding smoking rates for age groups
def impute_smoking_rate_source1(row):
    age_group = None
    if 15 <= row['age'] <= 17:
        age_group = '15–17'
    elif 18.0 <= row['age'] <= 24.0:
        age_group = '18–24'
    elif 25.0 <= row['age'] <= 34.0:
        age_group = '25–34'
    elif 35.0 <= row['age'] <= 44.0:
        age_group = '35–44'
    elif 45.0 <= row['age'] <= 54.0:
        age_group = '45–54'
    elif 55.0 <= row['age'] <= 64.0:
        age_group = '55–64'
    elif 65 <= row['age'] <= 74:
        age_group = '65–74'
    elif row['age'] >= 75:
        age_group = '75+'

    if age_group is None:
        return 0

    if row['sex'] == '0':  # Female
        return smoking_rates_source1[age_group]['Female']
    else:  # Male
        return smoking_rates_source1[age_group]['Male']

# For Source 2, apply the formula for males
def impute_smoking_rate_source2(row):
    age_group = None
    if 18 <= row['age'] <= 24:
        age_group = '18–24'
    elif 25 <= row['age'] <= 44:
        age_group = '25–44'
    elif 45 <= row['age'] <= 64:
        age_group = '45–64'
    elif row['age'] >= 65:
        age_group = '65+'

    if age_group is None:
        return 0

    if row['sex'] == '0':  # Female
        return smoking_rates_source2[age_group]['Female']
    else:  # Male
        male_rate = smoking_rates_source2[age_group]['Male']
        female_rate = smoking_rates_source2[age_group]['Female']
        return male_rate * (source2_smoking_rate_among_men / source2_smoking_rate_among_women)

# Apply the imputation for Source 1 and Source 2
df_cleaned['smoke_imputed_source1'] = df_cleaned.apply(impute_smoking_rate_source1, axis=1)
df_cleaned['smoke_imputed_source2'] = df_cleaned.apply(impute_smoking_rate_source2, axis=1)

# Normalize the imputed columns
df_cleaned['smoke_imputed_source1_normalized'] = (df_cleaned['smoke_imputed_source1'] - df_cleaned['smoke_imputed_source1'].min()) / (df_cleaned['smoke_imputed_source1'].max() - df_cleaned['smoke_imputed_source1'].min())
df_cleaned['smoke_imputed_source2_normalized'] = (df_cleaned['smoke_imputed_source2'] - df_cleaned['smoke_imputed_source2'].min()) / (df_cleaned['smoke_imputed_source2'].max() - df_cleaned['smoke_imputed_source2'].min())

print(df_cleaned.head())

    age  sex  painloc  painexer   cp  trestbps  smoke  fbs  prop  nitr  ...  \
0  63.0  1.0      1.0       1.0  1.0     145.0    NaN  1.0   0.0   0.0  ...   
1  67.0  1.0      1.0       1.0  4.0     160.0    NaN  0.0   1.0   0.0  ...   
2  67.0  1.0      1.0       1.0  4.0     120.0    NaN  0.0   1.0   0.0  ...   
3  37.0  1.0      1.0       1.0  3.0     130.0    NaN  0.0   1.0   0.0  ...   
4  41.0  0.0      1.0       1.0  2.0     130.0    NaN  0.0   0.0   0.0  ...   

   diuretic  thaldur  thalach  exang  oldpeak  slope  smoke_imputed_source1  \
0       0.0     10.5    150.0    0.0      2.3    3.0                   17.4   
1       0.0      9.5    108.0    1.0      1.5    2.0                    9.9   
2       0.0      8.5    129.0    1.0      2.6    2.0                    9.9   
3       0.0     13.0    187.0    0.0      3.5    3.0                   13.5   
4       0.0      7.0    172.0    0.0      1.4    1.0                   13.5   

   smoke_imputed_source2  smoke_imputed_source1_no

In [27]:
from sklearn.model_selection import train_test_split

In [30]:
# Task 3

# Features (X) are the columns in df_cleaned
X = df_cleaned  # Features

# Target (y) is the 'target' column from the original df
y = df['target']  # Target labels

# Drop rows where the target (y) is NaN
X = X[y.notna()]  # Keep only the rows where y is not NaN
y = y.dropna()  # Remove NaN values from y

# Perform stratified split (90-10 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, stratify=y, random_state=42)

In [36]:
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV, StratifiedKFold, cross_val_score
from sklearn.metrics import classification_report, roc_auc_score
import numpy as np

In [34]:
# Impute missing values using the median
imputer = SimpleImputer(strategy='median') 
X_imputed = imputer.fit_transform(X)  # Apply imputation to X

# train-test split and model training
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.1, stratify=y, random_state=42)

In [37]:
#Task 4

# Scale the data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)  # Apply scaling to training data
X_test_scaled = scaler.transform(X_test)  # Apply the same scaling to test data

# Define models
models = {
    'Logistic Regression': LogisticRegression(max_iter=1000),  # Increase max_iter to 1000
    'Random Forest': RandomForestClassifier(),
    'SVM': SVC(),
    'KNN': KNeighborsClassifier()
}

# Hyperparameter tuning
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [10, 20, None]
}
param_grid_lr = {
    'C': [0.1, 1, 10],
    'solver': ['liblinear', 'saga']
}

# 5-fold cross-validation setup
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# For each model, perform 5-fold cross-validation and hyperparameter tuning
results = {}
for model_name, model in models.items():
    print(f"Training {model_name}")
    
    if model_name == 'Random Forest':
        grid_search = GridSearchCV(estimator=model, param_grid=param_grid_rf, cv=cv, scoring='accuracy')
        grid_search.fit(X_train_scaled, y_train)  # Use scaled data for training
        best_model = grid_search.best_estimator_
    elif model_name == 'Logistic Regression':
        grid_search = GridSearchCV(estimator=model, param_grid=param_grid_lr, cv=cv, scoring='accuracy')
        grid_search.fit(X_train_scaled, y_train)  # Use scaled data for training
        best_model = grid_search.best_estimator_
    else:
        best_model = model  # No hyperparameter tuning for SVM and KNN for simplicity
        
    # 5-fold cross-validation
    cv_results = cross_val_score(best_model, X_train_scaled, y_train, cv=cv, scoring='accuracy')
    results[model_name] = cv_results

# Reporting performance
for model_name, cv_results in results.items():
    print(f"{model_name} cross-validation results:")
    print(f"Accuracy: Mean={cv_results.mean():.4f}, Std={cv_results.std():.4f}")

    # Train on the full dataset and assess performance
    best_model.fit(X_train_scaled, y_train)  # Train using the scaled data
    y_pred = best_model.predict(X_test_scaled)  # Make predictions on the scaled test data
    
    # Print classification report (Precision, Recall, F1-score)
    print(classification_report(y_test, y_pred))
    
    # Calculate and print ROC AUC
    if hasattr(best_model, "predict_proba"):  # Check if the model supports predict_proba method
        roc_auc = roc_auc_score(y_test, best_model.predict_proba(X_test_scaled)[:, 1])
    else:
        roc_auc = "Not available"  # Some models like SVM don't support predict_proba
    print(f"ROC AUC: {roc_auc}\n")

Training Logistic Regression
Training Random Forest
Training SVM
Training KNN
Logistic Regression cross-validation results:
Accuracy: Mean=0.8084, Std=0.0319
              precision    recall  f1-score   support

         0.0       0.75      0.60      0.67        40
         1.0       0.72      0.84      0.78        50

    accuracy                           0.73        90
   macro avg       0.74      0.72      0.72        90
weighted avg       0.74      0.73      0.73        90

ROC AUC: 0.7947499999999998

Random Forest cross-validation results:
Accuracy: Mean=0.8060, Std=0.0264
              precision    recall  f1-score   support

         0.0       0.75      0.60      0.67        40
         1.0       0.72      0.84      0.78        50

    accuracy                           0.73        90
   macro avg       0.74      0.72      0.72        90
weighted avg       0.74      0.73      0.73        90

ROC AUC: 0.7947499999999998

SVM cross-validation results:
Accuracy: Mean=0.7960, Std

Key Performance Metrics:
- Accuracy: The proportion of correctly predicted labels.
- Precision: The proportion of positive predictions that were actually correct.
- Recall: The proportion of actual positives that were correctly identified.
- F1-score: The harmonic mean of precision and recall.
- ROC AUC: The area under the Receiver Operating Characteristic curve, which provides an aggregate measure of performance across all possible classification thresholds.

Summary of Results

- Logistic Regression, Random Forest, SVM, and KNN all show similar cross-validation accuracy, around 80% for training, with slight variations.

Performance Metrics (Precision, Recall, F1-Score):

- Precision: All models have similar precision for both classes, around 0.72–0.75.
- Recall: Recall for the positive class (1.0) is around 0.84, which is quite good.
- F1-Score: The F1-score, which balances both precision and recall, is also consistent across models, around 0.73–0.78.
- ROC AUC: All models report an ROC AUC of ~0.795, indicating a good level of model performance. ROC AUC is particularly valuable because it assesses model performance across all possible thresholds, making it a good indicator of a model’s discriminatory ability.

Recommendations:

- Logistic Regression and Random Forest are performing quite similarly in terms of accuracy, precision, recall, and F1-score, so you could consider focusing on one of these models depending on other factors (e.g., interpretability, speed).
- SVM and KNN perform slightly worse in terms of accuracy, though the difference is not drastic.
- ROC AUC of ~0.795 is relatively strong, suggesting that the models can distinguish between the two classes fairly well.


Task 5 (Final Model Selection)

Given that your main goal is to achieve good performance in detecting heart disease (where detecting positive cases is crucial), Random Forest strikes a good balance between performance and interpretability (through feature importance). It is well-suited for this kind of task, and its robustness to overfitting and ability to capture complex relationships in the data makes it a good model. It does well in terms of accuracy, precision, recall, and F1-score and the ROC AUC of around 0.795 shows that the model has a strong ability to distinguish between the classes, which is important in medical applications like heart disease prediction.
